docs/vmalert: clarify available integrations (#10996)
VictoriaMetrics/VictoriaMetrics: docs/victoriametrics/vmalert.md

and [alerting](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/) rules is very
similar to Prometheus rules and configured using YAML. Configuration examples may be found
in [testdata](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/app/vmalert/config/testdata) folder.

Every `rule` belongs to a `group` and every configuration file may contain arbitrary number of groups:

```yaml
groups:
  [ - <rule_group> ]
```

> ### Groups

Reviewer: Btw, maybe it should be `## Groups` instead of `### Groups`? I don’t see why it belongs under `## QuickStart`.

```suggestion
## Groups
```

### Rules

> Every rule contains `expr` field for the expression to evaluate against configured datasource.

Reviewer 2: ```suggestion
Every rule contains an `expr` field for the expression to evaluate against the configured datasource.
```

# Labels to add or overwrite labels from other external label sources, such as group labels, before storing the result.
#
# In case of conflicts, original labels are kept with prefix `exported_`.
# As a special case, specifying a label with an empty string value removes the label from the result if it exists 
# in the original query result; otherwise, it is ignored.
#
# Labels do not support templating in https://docs.victoriametrics.com/victoriametrics/vmalert/#templating due to cardinality concerns. See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/8171.
labels:
  [ <labelname>: <labelvalue> ]

# Whether to print debug information into logs.
# Information includes requests sent to the datasource.
# information - it will be printed to logs.
# Logs are printed with INFO level, so make sure that -loggerLevel=INFO to see the output.
[ debug: <bool> | default = false ]

# Defines the number of rule's updates entries stored in memory
# and available for view on rule's Details page.
# Overrides `rule.updateEntriesLimit` value for this specific rule.
[ update_entries_limit: <integer> | default 0 ]
```

For recording rules to work `-remoteWrite.url` must be specified.

> ## Integrations

Reviewer: How about putting the `## Integrations` part after the `## Topology examples`?

Author: Moved in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/10996/commits/822b74b9ea258d510c561bce07a6cf59a3f08e79

Do you think it is worth to give more unique names to the anchors? For example, it could be `/victoriametrics/vmalert/#integrating-with-prometheus` instead of `/victoriametrics/vmalert/#prometheus`

Reviewer: I don’t think it is needed, since sections like `/victoriametrics/vmalert/#prometheus` don’t contain detailed integration information for Prometheus, nor do they need to. All of these sections are short descriptions, in most cases the section itself does not need to be referenced specifically, it is more for informing users that it is compatible.
For VictoriaLogs integration, users need to be aware of more detailed configuration, and in that case we will just refer them directly to https://docs.victoriametrics.com/victorialogs/vmalert/.

record: <string>

# The expression to evaluate. The expression language depends on the type value.
# By default, PromQL/MetricsQL expression is used. Other available types are "graphite" and "vlogs".
# See https://docs.victoriametrics.com/victoriametrics/vmalert/#integrations
expr: <string>

# Labels to add or overwrite labels from other external label sources, such as group labels, before storing the result.
#
# In case of conflicts, original labels are kept with prefix `exported_`.
# As a special case, specifying a label with an empty string value removes the label from the result if it exists 
# in the original query result; otherwise, it is ignored.
#
# Labels do not support templating in https://docs.victoriametrics.com/victoriametrics/vmalert/#templating due to cardinality concerns. See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/8171.
labels:
  [ <labelname>: <labelvalue> ]

# Whether to print debug information into logs.
# Information includes requests sent to the datasource.
# information - it will be printed to logs.
# Logs are printed with INFO level, so make sure that -loggerLevel=INFO to see the output.
[ debug: <bool> | default = false ]

> # Defines the number of rule's updates entries stored in memory

Reviewer 2: ```suggestion
# Defines the number of rule updates entries stored in memory
```

> # and available for view on rule's Details page.

Reviewer 2: ```suggestion
# and available for view on rule Details page.
```

field. See [how we use them](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/9751ea10983d42068487624849cac7ad6fd7e1d8/deployment/docker/rules/alerts-cluster.yml#L44)
to link alerting rule and the corresponding panel on Grafana dashboard.

## Multitenancy

> There are the following approaches exist for alerting and recording rules across

Reviewer: Maybe add a note that the approach is only available in VictoriaMetrics. For VictoriaLogs, currently it can be done via https://docs.victoriametrics.com/victorialogs/vmalert/#how-to-use-multitenancy-in-rules.

Author: Thanks! Updated in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/10996/commits/cb458a9d5dd3399d020e144d5f0c2d5620a9e67c

[data persisting when storage is unreachable](https://docs.victoriametrics.com/victoriametrics/vmagent/#replication-and-high-availability),
or time series modification via [relabeling](https://docs.victoriametrics.com/victoriametrics/relabeling/).

## Integrations

vmalert can be integrated with different data sources for alerting and recording rules. But it deliberately allows
configuring only one `datasource.url`. We recommend running separate instances of vmalert for each datasource type

> with specified `-rule.defaultRuleType=<datasource_type>` command-line flag.

Reviewer 2: ```suggestion
with the specified `-rule.defaultRuleType=<datasource_type>` command-line flag.
```

###### VictoriaMetrics

vmalert natively integrates with [VictoriaMetrics](https://docs.victoriametrics.com/victoriametrics/) for alerting and
recording rules.

###### VictoriaLogs

vmalert integrates with [VictoriaLogs](https://docs.victoriametrics.com/victorialogs/) and allows configuring alerting and recording rules using [LogsQL](https://docs.victoriametrics.com/victorialogs/logsql/).
Results of recording rules and alerting state should be persisted to the remote-write compatible storage, such as VictoriaMetrics.

> To enable VictoriaLogs compatibility set `-rule.defaultRuleType=vlogs` commmand-line flag.

Reviewer 2: ```suggestion
To enable VictoriaLogs compatibility set the `-rule.defaultRuleType=vlogs` command-line flag.
```

See [this doc](https://docs.victoriametrics.com/victorialogs/vmalert/) for details.

###### VictoriaTraces

vmalert integrates with [VictoriaTraces](https://docs.victoriametrics.com/victoriatraces/) in exactly the same way as
with [VictoriaLogs](https://docs.victoriametrics.com/victoriametrics/vmalert/#victorialogs).

###### Graphite

vmalert integrates with [Graphite Render API](https://graphite.readthedocs.io/en/stable/render_api.html) and allows configuring alerting and recording rules.
During evaluation, vmalert will send requests to `<-datasource.url>/render?format=json`.

> To enable Graphite compatibility set `-rule.defaultRuleType=graphite` commmand-line flag.

Reviewer 2: ```suggestion
To enable Graphite compatibility set the `-rule.defaultRuleType=graphite` command-line flag.
```

Since VictoriaMetrics supports both Graphite and Prometheus APIs, it is possible to mix Graphite and VictoriaMetrics rules.

> On the group level, set `type` field to specify to which datasource type it should belong: `prometheus` (MetricsQL) or `graphite` (GraphiteQL).

Reviewer 2: ```suggestion
On the group level, set the `type` field to specify to which datasource type it should belong: `prometheus` (MetricsQL) or `graphite` (GraphiteQL).
```

> When using vmalert with both `graphite` and `prometheus` rules configured against cluster version of VictoriaMetrics dont forget

Reviewer 2: ```suggestion
When using vmalert with both `graphite` and `prometheus` rules configured against the cluster version of VictoriaMetrics. Don't forget
```

> to set `-datasource.appendTypePrefix` flag to `true`, so vmalert can adjust URL prefix automatically based on the query type.

Reviewer 2: ```suggestion
to set the `-datasource.appendTypePrefix` flag to `true`, so vmalert can adjust URL prefix automatically based on the query type.
```

###### Prometheus

vmalert uses [Prometheus HTTP API](https://prometheus.io/docs/prometheus/latest/querying/api/#http-api) for querying
and [Prometheus Remote Write v1 protocol](https://prometheus.io/docs/specs/prw/remote_write_spec/) for persisting

> recording rules results and alerting state. Hence, it can be integrated with any Prometheus compatible storage

Reviewer 2: ```suggestion
recording rules results and alerting state. Hence, it can be integrated with any Prometheus-compatible storage
```

* [How to query vmalert from single-node VictoriaMetrics](https://docs.victoriametrics.com/victoriametrics/single-server-victoriametrics/#vmalert)
* [How to query vmalert from VictoriaMetrics cluster](https://docs.victoriametrics.com/victoriametrics/cluster-victoriametrics/#vmalert)

## Rules backfilling

vmalert supports alerting and recording rules backfilling (aka `replay`). In replay mode vmalert
can read the same rules configuration as normal, evaluate them on the given time range and backfill

> results via remote write to the configured storage. vmalert supports only `prometheus` datasource type for backfilling.

Reviewer 2: ```suggestion
results via remote write to the configured storage. vmalert supports only the `prometheus` datasource type for backfilling.
```

vmalert automatically disables caching on VictoriaMetrics side by sending `nocache=1` param. It allows
to prevent cache pollution and unwanted time range boundaries adjustment during backfilling.

> Results of recording rules `replay` should match with results of normal rules evaluation.

Reviewer 2: ```suggestion
Results of recording rules `replay` should match the results of normal rules evaluation.
```

> Results of alerting rules `replay` are time series reflecting [alert's state](#alerts-state-on-restarts).

Reviewer 2: ```suggestion
Results of alerting rules `replay` are the time series reflecting the [state of the alert](#alerts-state-on-restarts).
```
